pennylane/operation.py

```diff
@@ -300,48 +300,149 @@ def hash(self):
+        operator's wires in the wire order of the full vector space.
+        Otherwise, the defaults ``wires=[0,...,n]`` and ``wire_order=[0,...,n]``
+        are used, where ``n`` is the number of wires of this operator.
 
         **Example:**
 
-        >>> qml.RY._matrix(0.5)
-        >>> array([[ 0.96891242+0.j, -0.24740396+0.j],
-                   [ 0.24740396+0.j,  0.96891242+0.j]])
+        >>> qml.RY.compute_matrix(0.5)
+        array([[ 0.96891242, -0.24740396],
+               [ 0.24740396,  0.96891242]])
+
+        >>> qml.RY.compute_matrix(0.5, wires=["b"], wire_order=["a", "b"])
+        tensor([[ 0.96891242, -0.24740396, 0., 0.],
+                [ 0.24740396,  0.96891242, 0., 0.],
+                [ 0., 0., 0.96891242, -0.24740396],
+                [ 0., 0., 0.24740396,  0.96891242]
+                ])
+
+        >>> qml.RY.compute_matrix(0.5, wires=["b"], wire_order=["b", "a"])
+        tensor([[ 0.96891242, 0., -0.24740396, 0.],
+                [0.,  0.96891242, 0., -0.24740396],
+                [ 0.24740396, 0., 0.96891242, 0.],
+                [0.,  0.24740396, 0., 0.96891242+0.j],
+                ])
+
+        Args:
+            params (Iterable): trainable parameters that influence the matrix representation
+            wires (Iterable): wires of this operator
+            wire_order (Iterable): global wire order, must contain all wire labels in ``wires``
+            hyperparams (dict): other inputs that influence the matrix representation
 
         Returns:
-            array: matrix representation
+            tensor-like: matrix representation, or None if not defined
         """
-        raise NotImplementedError
+        base_matrix = cls._matrix(*params, **hyperparams)
+        interface = qml.math._multi_dispatch(params)  # pylint: disable=protected-access
 
-    @property
-    def matrix(self):
+        # representation is not defined
+        if base_matrix is None:
+            return None
+
+        # cannot have wire_order but no wires to interpret
+        if wires is None and wire_order is not None:
+            raise ValueError("Expected wires when wire order is defined, got None.")
+
+        # intersecting this case here because it is common
+        # and we want to avoid casting to Wires unless necessary
+        if wire_order is None:
+            # TODO: cast to interface?
+            return base_matrix
+
+        wire_order = Wires(wire_order)
+        wires = Wires(wires)
+
+        # base matrix requested
+        if Wires(wires) == Wires(wire_order):
+            # TODO: cast to interface?
+            return base_matrix
+
+        n = len(wires)
+
+        # operator's wire positions relative to wire ordering
+        op_wire_pos = Wires(wire_order).indices(wires)
+
+        I = qml.math.reshape(
+            qml.math.eye(2 ** len(wire_order), like=interface), [2] * len(wire_order) * 2
+        )
+        axes = (list(range(n, 2 * n)), op_wire_pos)
```

Comment:
I realize, from reading this: this is important internal logic that a user should _not_ overwrite or change!

As a result, it maybe doesn't make sense for a user to overwrite `compute_matrix()`.

Would it make sense to:

- Have a staticmethod `compute_matrix(parameters, hyperparameters)` that _simply returns the base matrix representation_.

- Have an instance method `matrix(self, wire_order)` that includes this wire ordering/expansion logic?

Reply:
I was thinking that code for rearranging the base matrix to have a different order could be extracted to a utility method, or even made a "transform".  Then this method would be much simpler.

Reply:
Good idea!

Reply:
Here is the problem @josh146: if the expansion logic is in `matrix()` only, we cannot cache it. Otherwise I'd be happy to merge the two! 

Maybe I can benchmark expansion and see how costly it is.

Reply:
From our phone call just now: caching will be preferred on a staticmethod rather than a classmethod!